### PR TITLE
Refactored compaction_session.py

### PR DESCRIPTION
**Pull request by akindu-amazon**
Responding to comments from https://github.com/ray-project/deltacat/pull/333

### CODE REVIEW

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (DmitriGekhtman) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:84]: Method name could perhaps be more descriptive --
We're not executing the compaction here, so maybe `_fetch_compaction_metadata` or `_prepare_for_compaction`.

We could also use either a docstring for this method or more explanation at this method's call site in compaction_session.

We should also add type annotations for the returned tuple.
You could consider defining a dataclass for the returned tuple, but it might be unnecessary.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (DmitriGekhtman) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:104]: The caller of this method already has access to the `params`, so it might not make sense to access and return `task_max_parallelism` here -- it should be possible to extract the max_parallelism later in compaction_session when you need it.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (DmitriGekhtman) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:157]: Is the `audit_url` already present in `mutable_compaction_audit`?
If so, might not be necessary to return `audit_url` as a separate parameter.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (DmitriGekhtman) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:154]: I haven't looked carefully at the details, but it seems there are only three (mostly) independent pieces of data to return here:
`mutable_compacton_audit`, `previous_compacted_delta_manifest`,  and `round_completion_info`.

**Review by DmitriGekhtman — COMMENTED** [review]
Thanks for cleaning up the super long compaction function.

Haven't gone through the entire diff yet, but added some comments on `_start_execute_compaction`

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (DmitriGekhtman) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:207]: I think the Any's are `float`, `int`, `int`, `float`.

Do you think it might be possible to simplify the output of `_run_merge` and correspondingly the input args of `_complete_round`?

Also, perhaps you could use a more descriptive name for `_complete_round`. 

Anywhere else you can think of ways to simplify the inputs and outputs of these compaction utils, that'd help.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (DmitriGekhtman) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:548]: Seems this s3 upload happens in both `_process_merge_results` and `_get_compaction_result`?
Is that intended?

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (Zyiqin-Miranda) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:677]: Minor: Could we rename this method to `commit_compaction_result`?

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (Zyiqin-Miranda) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:713]: Could we make this function to just `commit_partition` purpose and not returning a `round_completion_files_s3_url`?

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (Zyiqin-Miranda) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:566]: Minor: Could we rename this method to `upload_compaction_audit` and just upload audit in this function?

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (Zyiqin-Miranda) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:626]: We can put this remaining function into a new function like `write_new_round_completion_file`?

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (Zyiqin-Miranda) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:170]: Could we put this creating `compacted_stream_locator` into another helper function like `_create_compacted_partition` since it's not building uniform deltas?

**Review by Zyiqin-Miranda — COMMENTED** [review]
Looks good overall! Thanks for the refactoring!

**Review by DmitriGekhtman — APPROVED** [review]
Thanks!
I'd suggest getting a review from the subject matter experts (Miranda and/or Raghavendra) before merging.

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:85]: Definitely add one test to check if all audit info is being set.

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:110]: Didn't we decide to move this into CompactPartitionParams?

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:147]: Why is _build_uniform_deltas responsible for creating a compacted partition? I believe it's a separate concern.

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:162]: Didn't we discuss getting rid of this arg and get it from params?

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:170]: Didn't we discuss update this in audit where this is calculated?

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/compaction_session.py:178]: This log can be moved inside where we are generating this data.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:77]: utils generally contain public methods which can be used across the codebase. This is a private method which we don't intend to be used by anyone apart from execute_compaction method. Hence, can we move this to `/private/compaction_session_utils.py` module?

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:403]: Same here regarding task_max_parallelism.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:192]: This method runs both hash bucketing and merge but the method name says it only runs merge.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:399]: The order for methods is not followed. To reduce scroll up and down, keep methods used in methods in topological order.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:213]: options provider and input provider both must be in the same function as they share the same concern i.e., argument to hash bucket.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [inline deltacat/compute/compactor_v2/utils/compaction_utils.py:456]: This is the correct placement. You have merge options provider inside merge. Follow the same pattern for hash bucket.

**Review by raghumdani — CHANGES_REQUESTED** [review]
Thanks for refactoring. Please address few comments.

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [reply deltacat/compute/compactor_v2/compaction_session.py:87]: Any update?

Comment (raghumdani) [reply deltacat/compute/compactor_v2/compaction_session.py:87]: Yes, that's correct. You can check what properties are set by skimming through the code.

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [reply deltacat/compute/compactor_v2/compaction_session.py:112]: Yes, should work. nit: It should be self.pg_config

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [reply deltacat/compute/compactor_v2/compaction_session.py:176]: Got it.. That explains the reasoning behind returning telemetry_time_merge but what about total_dd_record_count and other?

File: deltacat/compute/compactor_v2/compaction_session.py
Comment (raghumdani) [reply deltacat/compute/compactor_v2/compaction_session.py:185]: It's ok to split.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [reply deltacat/compute/compactor_v2/utils/compaction_utils.py:77]: There is no private dir yet. You might have to create one.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [reply deltacat/compute/compactor_v2/utils/compaction_utils.py:176]: Yes, that works.

File: deltacat/compute/compactor_v2/utils/compaction_utils.py
Comment (raghumdani) [reply deltacat/compute/compactor_v2/utils/compaction_utils.py:389]: The above ordering is correct. Just make sure it's reverse.

File: deltacat/compute/compactor_v2/private/compaction_utils.py
Comment (Zyiqin-Miranda) [inline deltacat/compute/compactor_v2/private/compaction_utils.py:717]: Unnecessary return?

File: deltacat/compute/compactor_v2/utils/task_options.py
Comment (Zyiqin-Miranda) [inline deltacat/compute/compactor_v2/utils/task_options.py:148]: Debug print statement?

**Review by Zyiqin-Miranda — APPROVED** [review]

File: deltacat/compute/compactor/model/compact_partition_params.py
Comment (raghumdani) [inline deltacat/compute/compactor/model/compact_partition_params.py:189]:
```suggestion
            cluster_resources = self.pg_config.resource
```

File: deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py
Comment (raghumdani) [inline deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py:690]: Why is this one None? Do we want to have a separate assert?

File: deltacat/tests/compute/test_compact_partition_incremental.py
Comment (raghumdani) [inline deltacat/tests/compute/test_compact_partition_incremental.py:216]: Not used anywhere in this method?

File: deltacat/tests/compute/test_compact_partition_rebase.py
Comment (raghumdani) [inline deltacat/tests/compute/test_compact_partition_rebase.py:198]: Not used anywhere in this method?

File: deltacat/tests/compute/test_util_common.py
Comment (raghumdani) [inline deltacat/tests/compute/test_util_common.py:227]: These are not definitely all attributes. Look at the actual audit file and ensure every one among them is there.

**Review by raghumdani — COMMENTED** [review]
Overall, looks good. Please address few minor comments.

File: deltacat/tests/compute/test_compact_partition_incremental.py
Comment (akindu-amazon) [reply deltacat/tests/compute/test_compact_partition_incremental.py:216]: If I don't include these parameters, I get an error when calling pytest.mark.parametrize. I think they are needed here even though the linter grays them out. For example, the `test_name` parameter is also not used in the method `test_compact_partition_incremental` but is needed during parametrizing. I intended to only add these new arguments to `rebase_then_incremental`, but because I added `assert_compaction_audit` to the `BaseCompactorTestCase` for simplicity and `compactor_version` during enriching (all three test case files use the same function to enrich), `test_compact_partition_incremental` and `test_compact_partition_rebase` need the args as well.

File: deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py
Comment (akindu-amazon) [reply deltacat/tests/compute/compact_partition_rebase_then_incremental_test_cases.py:690]: My mistake, this one is no longer None.

File: deltacat/tests/compute/test_util_common.py
Comment (akindu-amazon) [reply deltacat/tests/compute/test_util_common.py:227]: These assertions are based on what `compaction_audit` updates are made in the refactored `compaction_session` and `compaction_utils`:
```
{'deltacatVersion': '1.1.11', 
'rayVersion': '2.32.0', 
'auditUrl': 's3://test-compaction-artifacts-bucket/compaction-audit/dadad0e699a014b0db94016217c90c63ebd81d80.json', 
'hashBucketCount': 3, 
'compactorVersion': 'V2', 
'inputSizeBytes': 6461, 
'inputFileCount': 4, 
'estimatedInMemorySizeBytesDuringDiscovery': 280, 
'uniformDeltasCreated': 1, 
'deltaDiscoveryTimeInSeconds': 0.0067308819852769375, 
'hashBucketTimeInSeconds': 0.01295870600733906, 
'hashBucketInvokeTimeInSeconds': 0.012178680044598877, 
'hashBucketResultSize': 464, 
'totalClusterObjectStoreMemoryBytes': 4091910144.0, 
'totalClusterMemoryBytes': 8183820288.0, 
'totalObjectStoreMemoryUsedBytes': 0.0, 
'hashBucketPostObjectStoreMemoryUsedBytes': 0.0, 
'hashBucketResultWaitTimeInSeconds': 0.001363515853881836, 
'hashBucketTaskPeakMemoryUsedBytes': 314687488.0, 
'hashBucketProcessedSizeBytes': 284, 
'inputRecords': 20, 
'mergeTimeInSeconds': 0.04835349402856082, 
'mergeInvokeTimeInSeconds': 0.04747437103651464, 
'mergeResultSize': 29610, 
'mergePostObjectStoreMemoryUsedBytes': 0.0, 
'mergeResultWaitTimeInSeconds': 1406994735.0460403, 
'mergeTaskPeakMemoryUsedBytes': 314687488.0, 
'recordsDeduped': 10, 
'recordsDeleted': 0, 
'compactionTimeInSeconds': 0.5707382939290255, 
'peakMemoryUsedBytesCompactionSessionProcess': 314687488, 
'untouchedFileCount': 0, 
'untouchedFileRatio': 0.0, 
'untouchedRecordCount': 0, 
'untouchedSizeBytes': 0, 
'outputFileCount': 3, 
'outputSizeBytes': 4760, 
'outputSizePyarrowBytes': 292, 
'peakMemoryUsedBytesPerTask': 314687488.0, 
'pyarrowVersion': '12.0.1', 
'telemetryTimeInSeconds': 0.001723743975162506}
```
This is a sample one from one of the `rebase_then_incremental` test cases. Should I test for all of these values? My  main worry is that values that are 0 or 0.0 would fail `not None` assertions.

File: deltacat/tests/compute/test_compact_partition_incremental.py
Comment (raghumdani) [reply deltacat/tests/compute/test_compact_partition_incremental.py:216]: Yeah, you can actually add the assertion here.

File: deltacat/tests/compute/test_util_common.py
Comment (raghumdani) [reply deltacat/tests/compute/test_util_common.py:227]: Yes please. 0.0 won't fail not None.

**Review by raghumdani — APPROVED** [review]
LGTM

**Pull request merged**
